### PR TITLE
Scripting: Deprecate general cache settings

### PR DESCRIPTION
**Pull request by stu-elastic**
Deprecate and remove as fallback settings:
* `script.cache.max_size` in favor of `script.context.$CONTEXT.cache_max_size`
* `script.cache.expire` in favor of `script.context.$CONTEXT.cache_expire`
* `script.max_compilations_rate` in favor of `script.context.$CONTEXT.max_compilations_rate`

Default to context cache.

Add `script.disable_max_compilations_rate`, default `false` to totally disable compilation rates.  This setting is intended for integration tests.

Refs: #50152

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 1]
```diff
@@ -688,25 +691,25 @@ ScriptCacheStats cacheStats() {
...
             }
             return new ScriptCacheStats(context);
         }
 
         /**
-         * Update settings for the context cache, if we're in the context cache mode otherwise no-op.
+         * Update a single context cache if we're in the context cache mode otherwise no-op.
          */
-        void updateContextSettings(Settings settings, ScriptContext<?> context) {
+        void set(String name, ScriptCache context) {
```
Comment (stu-elastic) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java:703]: Rename to `cache`

**Review by jdconrad — APPROVED** [review]
Thank you for the walkthrough. This looks good to me. Please also have @rjernst review as he's more familiar with the testing framework.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 2]
```diff
@@ -576,89 +583,85 @@ public void applyClusterState(ClusterChangedEvent event) {
...
 
-    /**
-     * Container for the ScriptCache(s).  This class operates in two modes:
-     * 1) general mode, if the general script cache is configured.  There are no context caches in this case.
-     * 2) context mode, if the context script cache is configured.  There is no general cache in this case.
-     */
-    static class CacheHolder {
-        final ScriptCache general;
-        final Map<String, AtomicReference<ScriptCache>> contextCache;
+    void setCacheHolder(Settings settings, AtomicReference<CacheHolder> cacheHolder) {
```
Comment (stu-elastic) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java:586]: Make this package protected.

File: docs/reference/modules/indices/circuit_breaker.asciidoc [hunk 3]
```diff
@@ -111,8 +111,8 @@ within a period of time.
 See the "prefer-parameters" section of the <<modules-scripting-using,scripting>>
 documentation for more information.
 
-`script.max_compilations_rate`::
+`script.context.$CONTEXT.max_compilations_rate`::
 
     Limit for the number of unique dynamic scripts within a certain interval
-    that are allowed to be compiled. Defaults to 75/5m, meaning 75 every 5 
-    minutes.
+    that are allowed to be compiled for a given context. Defaults to 75/5m,
```
Comment (jrodewig) [inline docs/reference/modules/indices/circuit_breaker.asciidoc:117]:
```suggestion
    that are allowed to be compiled for a given context. Defaults to `75/5m`,
```

File: docs/reference/scripting/using.asciidoc [hunk 4]
```diff
@@ -98,9 +98,11 @@ second version is only compiled once.
 
 If you compile too many unique scripts within a small amount of time,
 Elasticsearch will reject the new dynamic scripts with a
-`circuit_breaking_exception` error. By default, up to 15 inline scripts per
-minute will be compiled. You can change this setting dynamically by setting
-`script.max_compilations_rate`.
+`circuit_breaking_exception` error. By default, up to 75 scripts per
+5 minutes will be compiled for most contexts and 375 scripts per 5 minutes
+for ingest contexts.  You can change these settings dynamically by setting
```
Comment (jrodewig) [inline docs/reference/scripting/using.asciidoc:103]: We do single space between sentences.
```suggestion
for ingest contexts. You can change these settings dynamically by setting
```

**Review by jrodewig — APPROVED** [review]
Docs LGTM.

File: docs/reference/modules/indices/circuit_breaker.asciidoc (hunk 3 above)
Comment (stu-elastic) [reply docs/reference/modules/indices/circuit_breaker.asciidoc]: mono'd

File: docs/reference/scripting/using.asciidoc (hunk 4 above)
Comment (stu-elastic) [reply docs/reference/scripting/using.asciidoc]: single spaced.

**Review by rjernst — APPROVED** [review]
LGTM after https://github.com/elastic/elasticsearch/pull/55560 is merged.

**Comment by stu-elastic** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by stu-elastic** [comment]
Master: ef543b0
7.x (7.9): 88e8b34

**Pull request merged**